src/Advection/weno.jl

```diff
@@ -0,0 +1,125 @@
+
+@inline function weno_weights_x(i, j, k, weno::WENO{K}, ϕ) where K
+    α = [αx(i, j, k, weno, r, ϕ) for r in 0:K-1]
+    return α ./ sum(α)
+end
+
+@inline function weno_weights_y(i, j, k, weno::WENO{K}, ϕ) where K
+    α = [αy(i, j, k, weno, r, ϕ) for r in 0:K-1]
+    return α ./ sum(α)
+end
+
+@inline function weno_weights_z(i, j, k, weno::WENO{K}, ϕ) where K
+    α = [αz(i, j, k, weno, r, ϕ) for r in 0:K-1]
+    return α ./ sum(α)
+end
+
+#####
+##### WENO flux reconstruction
+#####
+
+@inline function weno_flux_x(i, j, k, weno::WENO{K}, ϕ) where K
+    w = weno_weights_x(i, j, k, weno, ϕ)
+    return sum(w[r+1] * px(i, j, k, weno, r, ϕ) for r in 0:K-1)
+end
+
+@inline function weno_flux_y(i, j, k, weno::WENO{K}, ϕ) where K
+    w = weno_weights_y(i, j, k, weno, ϕ)
+    return sum(w[r+1] * py(i, j, k, weno, r, ϕ) for r in 0:K-1)
+end
+
+@inline function weno_flux_z(i, j, k, weno::WENO{K}, ϕ) where K
+    w = weno_weights_z(i, j, k, weno, ϕ)
+    return sum(w[r+1] * pz(i, j, k, weno, r, ϕ) for r in 0:K-1)
+end
+
+#####
+##### Momentum advection fluxes
+#####
+
+@inline momentum_flux_uu(i, j, k, grid, weno::WENO, u)    = ℑxᶜᵃᵃ(i, j, k, grid, Ax_ψᵃᵃᶠ, u) * weno_flux_x(i, j, k, weno, u)
+@inline momentum_flux_uv(i, j, k, grid, weno::WENO, u, v) = ℑxᶠᵃᵃ(i, j, k, grid, Ay_ψᵃᵃᶠ, v) * weno_flux_y(i, j, k, weno, u)
+@inline momentum_flux_uw(i, j, k, grid, weno::WENO, u, w) = ℑxᶠᵃᵃ(i, j, k, grid, Az_ψᵃᵃᵃ, w) * weno_flux_z(i, j, k, weno, u)
+
+@inline momentum_flux_vu(i, j, k, grid, weno::WENO, u, v) = ℑyᵃᶠᵃ(i, j, k, grid, Ax_ψᵃᵃᶠ, u) * weno_flux_x(i, j, k, weno, v)
+@inline momentum_flux_vv(i, j, k, grid, weno::WENO, v)    = ℑyᵃᶜᵃ(i, j, k, grid, Ay_ψᵃᵃᶠ, v) * weno_flux_y(i, j, k, weno, v)
+@inline momentum_flux_vw(i, j, k, grid, weno::WENO, v, w) = ℑyᵃᶠᵃ(i, j, k, grid, Az_ψᵃᵃᵃ, w) * weno_flux_z(i, j, k, weno, v)
+
+@inline momentum_flux_wu(i, j, k, grid, weno::WENO, u, w) = ℑzᵃᵃᶠ(i, j, k, grid, Ax_ψᵃᵃᶠ, u) * weno_flux_x(i, j, k, weno, w)
+@inline momentum_flux_wv(i, j, k, grid, weno::WENO, v, w) = ℑzᵃᵃᶠ(i, j, k, grid, Ay_ψᵃᵃᶠ, v) * weno_flux_y(i, j, k, weno, w)
+@inline momentum_flux_ww(i, j, k, grid, weno::WENO, w)    = ℑzᵃᵃᶜ(i, j, k, grid, Az_ψᵃᵃᵃ, w) * weno_flux_z(i, j, k, weno, w)
+
+#####
+##### Advective tracer fluxes
+#####
+
+@inline advective_tracer_flux_x(i, j, k, grid, weno::WENO, u, c) = Ax_ψᵃᵃᶠ(i, j, k, grid, u) * weno_flux_x(i, j, k, weno, c)
+@inline advective_tracer_flux_y(i, j, k, grid, weno::WENO, v, c) = Ay_ψᵃᵃᶠ(i, j, k, grid, v) * weno_flux_y(i, j, k, weno, c)
+@inline advective_tracer_flux_z(i, j, k, grid, weno::WENO, w, c) = Az_ψᵃᵃᵃ(i, j, k, grid, w) * weno_flux_z(i, j, k, weno, c)
+
+#####
+##### Need to advect momentum like tracers
+#####
+
+@inline div_ũu(i, j, k, grid, advection::WENO, U) = div_uc(i, j, k, grid, advection, U, U.u)
+@inline div_ũv(i, j, k, grid, advection::WENO, U) = div_uc(i, j, k, grid, advection, U, U.v)
+@inline div_ũw(i, j, k, grid, advection::WENO, U) = div_uc(i, j, k, grid, advection, U, U.w)
```

Comment:
This is sketchy --- and might be why the thermal bubble test blows up. The indexing is off at the boundaries.

Reply:
I'm not sure we should merge until we can sort out this issue. I'm worried the convergence test does not uncover this issue. Note that because of the staggered grid indexing convention, this index shift may have different behavior for "forward" advection versus "backward" advection (in negative directions).

Reply:
Are there any references using WENO for momentum advection on a staggered grid that explain this?

Reply:
Ah --- the reason the convergence tests can't catch this error is because they test advection by a uniform velocity. Thus the convergence test does not catch an error that has to do with the cell interface to which the advecting velocity field is interpolated. That's the only difference between the two?

Reply:
Resolved by pair programming.